File: app/src/main/java/com/mypos/store/presentation/homeRefactor/view/HomeFragment.java
package com.mypos.store.presentation.homeRefactor.view;

import static androidx.core.os.BundleKt.bundleOf;

import android.content.Context;
import android.content.ContextWrapper;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.mypos.store.R;
import com.mypos.store.domain.articles.model.ArticleEntity;
import com.mypos.store.presentation.homeRefactor.model.HomeEventsCallback;
import com.mypos.store.presentation.homeRefactor.view.adapter.ArticlesAdapter;
import com.mypos.store.presentation.homeRefactor.view.adapter.ItemClickListener;
import com.mypos.store.presentation.homeRefactor.viewmodel.HomeViewModel;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class HomeFragment extends Fragment implements View.OnClickListener, HomeEventsCallback {

    private View view;
    private RecyclerView recyclerView;
    private ImageButton buttonCart, buttonAddNew;
    private TextView textViewNoProducts;
    private ImageView imageViewNoProducts;
    private ProgressBar progressBar;
    private ArticlesAdapter adapter;

    private HomeViewModel viewModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_home, container, false);
        init();
        viewModel = new ViewModelProvider(requireActivity()).get(HomeViewModel.class);
        progressBar.setVisibility(View.VISIBLE);
        viewModel.init(this);

        viewModel.readArticles();
        return view;
    }


    private void init() {
        buttonCart = view.findViewById(R.id.cartButton);
        recyclerView = view.findViewById(R.id.rvArticles);
        buttonAddNew = view.findViewById(R.id.addNewButton);
        textViewNoProducts = view.findViewById(R.id.textViewNoProducts);
        imageViewNoProducts = view.findViewById(R.id.imageViewNoProducts);
        progressBar = view.findViewById(R.id.progressBar);

        buttonCart.setOnClickListener(this);
        buttonAddNew.setOnClickListener(this);

        ContextWrapper cw = new ContextWrapper(requireContext());
        File directory = cw.getDir("articlesImages", Context.MODE_APPEND);

        adapter = new ArticlesAdapter(new ArrayList<ArticleEntity>(), new ItemClickListener() {
            @Override
            public void onArticleClick(int articleId) {
                openDetails(articleId);
            }

            @Override
            public void onCartActionClick(ArticleEntity article, boolean isIncreased) {
                viewModel.cartAction(article, isIncreased);
                onArticleUpdated(article);
            }
        }, directory.getAbsolutePath());

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.cartButton) {
            try {
                Navigation.findNavController(view).navigate(
                        R.id.action_refactoredHomeFragment_to_cartFragment
                );
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else if (v.getId() == R.id.addNewButton) {
            try {
                Navigation.findNavController(view).navigate(
                        R.id.action_refactoredHomeFragment_to_addNewFragment
                );
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onFetched(List<ArticleEntity> articles) {
        progressBar.setVisibility(View.GONE);
        if (articles.isEmpty()) {
            textViewNoProducts.setVisibility(View.VISIBLE);
            imageViewNoProducts.setVisibility(View.VISIBLE);
        }
        adapter.addMultipleItems(articles);
    }

    @Override
    public void onArticleAdded(ArticleEntity article) {
        adapter.addNewItem(article);

        if (adapter.getItemCount() == 1) {
            textViewNoProducts.setVisibility(View.GONE);
            imageViewNoProducts.setVisibility(View.GONE);
        }
    }

    @Override
    public void onArticleDeleted(ArticleEntity article) {
        adapter.deleteItem(article);

        if (adapter.getItemCount() < 1) {
            textViewNoProducts.setVisibility(View.VISIBLE);
            imageViewNoProducts.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onArticleUpdated(ArticleEntity article) {
        adapter.updateItem(article);
    }

    @Override
    public void onError(String message) {
        Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT).show();
    }

    public void openDetails(int id) {
        Bundle bundle = new Bundle();
        bundle.putInt("productId", id);
        Navigation.findNavController(view).navigate(
                R.id.action_refactoredHomeFragment_to_detailsFragment, bundle
        );
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/mypos/store/presentation/home/view/HomeFragment.kt
package com.mypos.store.presentation.home.view

import android.content.Context
import android.content.ContextWrapper
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.core.os.bundleOf
import androidx.fragment.app.viewModels
import androidx.navigation.Navigation
import com.mypos.store.R
import com.mypos.store.databinding.FragmentHomeBinding
import com.mypos.store.presentation.base.viewmodel.observeIn
import com.mypos.store.presentation.home.model.HomeModel
import com.mypos.store.presentation.home.viewmodel.HomeViewModel
import com.mypos.store.presentation.ui.buttons.AddNewButton
import com.mypos.store.presentation.ui.articles.ArticlesList
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.onEach

@AndroidEntryPoint
class HomeFragment : Fragment() {

    private val viewModel: HomeViewModel by viewModels()
    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        val view = binding.root


        viewModel.uiState
            .onEach(::onHandleState)
            .observeIn(this)

        viewModel.setEvent(HomeModel.HomeUiEvent.LoadArticles)

        initCompose()
        return view
    }

    private fun onHandleState(state: HomeModel.HomeUiState) {
        if (state.isLoading) {
            binding.progressBar.visibility = View.VISIBLE
        } else {
            binding.progressBar.visibility = View.GONE
        }
    }

    private fun initCompose() {
        binding.addNewButton.apply {
            setContent {
                Button(colors = ButtonDefaults.buttonColors(containerColor = Color.Black),
                    modifier = Modifier.size(50.dp),
                    shape = CircleShape,
                    contentPadding = PaddingValues(0.dp),  //avoid the little icon
                    onClick = {
                        try {
                            Navigation.findNavController(binding.root).navigate(
                                R.id.action_homeFragment_to_addNewFragment
                            )
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }) {
                    Icon(Icons.Default.Add, contentDescription = "add icon")
                }
            }
        }

        binding.cartButton.apply {
            setContent {
                Button(colors = ButtonDefaults.buttonColors(containerColor = Color.Black),
                    modifier = Modifier.size(50.dp),
                    shape = CircleShape,
                    contentPadding = PaddingValues(0.dp),  //avoid the little icon
                    onClick = {
                        try {
                            Navigation.findNavController(binding.root).navigate(
                                R.id.action_homeFragment_to_cartFragment
                            )
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }

                    }) {
                    Icon(Icons.Default.ShoppingCart, contentDescription = "cart icon")
                }
            }
        }
        val cw = ContextWrapper(context)
        val directory = cw.getDir("articlesImages", Context.MODE_APPEND)

        binding.articlesList.apply {
            setContent {
                val state = viewModel.uiState.collectAsState().value
                ArticlesList(state.articles, state.cart, { increase, id ->
                    viewModel.setEvent(HomeModel.HomeUiEvent.AddToCart(increase, id))
                }, onClickArticle = {
                    val bundle = bundleOf("productId" to it)

                    Navigation.findNavController(binding.root).navigate(
                        R.id.action_homeFragment_to_detailsFragment, bundle
                    )
                }, imagesDirPath = directory.absolutePath
                )
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }
}